core/src/Kokkos_HIP_Space.hpp

```diff
@@ -0,0 +1,633 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_HIPSPACE_HPP
+#define KOKKOS_HIPSPACE_HPP
+
+#include <Kokkos_Core_fwd.hpp>
+
+#if defined(KOKKOS_ENABLE_HIP)
+
+#include <iosfwd>
+#include <typeinfo>
+#include <string>
+
+#include <Kokkos_HostSpace.hpp>
+
+#include <hip/hip_runtime_api.h>
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+namespace Experimental {
+/** \brief  HIP on-device memory management */
+
+class HIPSpace {
+ public:
+  //! Tag this class as a kokkos memory space
+  typedef HIPSpace memory_space;
+  typedef Kokkos::Experimental::HIP execution_space;
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+
+  typedef unsigned int size_type;
+
+  /*--------------------------------*/
+
+  HIPSpace();
+  HIPSpace(HIPSpace&& rhs)      = default;
+  HIPSpace(const HIPSpace& rhs) = default;
+  HIPSpace& operator=(HIPSpace&& rhs) = default;
+  HIPSpace& operator=(const HIPSpace& rhs) = default;
+  ~HIPSpace()                              = default;
+
+  /**\brief  Allocate untracked memory in the rocm space */
+  void* allocate(const size_t arg_alloc_size) const;
+
+  /**\brief  Deallocate untracked memory in the rocm space */
+  void deallocate(void* const arg_alloc_ptr, const size_t arg_alloc_size) const;
+
+  /**\brief Return Name of the MemorySpace */
+  static constexpr const char* name() { return m_name; };
+
+  /*--------------------------------*/
+  /** \brief  Error reporting for HostSpace attempt to access HIPSpace */
+  static void access_error();
+  static void access_error(const void* const);
+
+ private:
+  int m_device;  ///< Which HIP device
+
+  static constexpr const char* m_name = "HIP";
+  friend class Kokkos::Impl::SharedAllocationRecord<
+      Kokkos::Experimental::HIPSpace, void>;
+};
+
+}  // namespace Experimental
+
+namespace Impl {
+
+/// \brief Initialize lock array for arbitrary size atomics.
+///
+/// Arbitrary atomics are implemented using a hash table of locks
+/// where the hash value is derived from the address of the
+/// object for which an atomic operation is performed.
+/// This function initializes the locks to zero (unset).
+void init_lock_arrays_rocm_space();
```

Comment:
even if those don't work yet: shouldn't those be named `init_lock_arrays_hip_space`?